# CRUD in MongoDB:

In this notebook we will explore the main transactions done on a database (CRUD) on a document based database (MongoDB).

In [65]:
import pymongo as pg #Among other things handles conncection between the db and python environement 
from pprint import pprint #pretty print
import datetime #Needed to format the datetime input later on.

In [66]:
client=pg.MongoClient()
db=client.test

In [67]:
print(db.restaurants.find_one())

{'_id': ObjectId('5cc7253f3ff99b33afc769ee'), 'address': {'building': '1007', 'coord': [-73.856077, 40.848447], 'street': 'Morris Park Ave', 'zipcode': '10462'}, 'borough': 'Bronx', 'cuisine': 'Bakery', 'grades': [{'date': datetime.datetime(2014, 3, 3, 0, 0), 'grade': 'A', 'score': 2}, {'date': datetime.datetime(2013, 9, 11, 0, 0), 'grade': 'A', 'score': 6}, {'date': datetime.datetime(2013, 1, 24, 0, 0), 'grade': 'A', 'score': 10}, {'date': datetime.datetime(2011, 11, 23, 0, 0), 'grade': 'A', 'score': 9}, {'date': datetime.datetime(2011, 3, 10, 0, 0), 'grade': 'B', 'score': 14}], 'name': 'Morris Park Bake Shop', 'restaurant_id': '30075445'}


In [68]:
pprint(db.restaurants.find_one())

{'_id': ObjectId('5cc7253f3ff99b33afc769ee'),
 'address': {'building': '1007',
             'coord': [-73.856077, 40.848447],
             'street': 'Morris Park Ave',
             'zipcode': '10462'},
 'borough': 'Bronx',
 'cuisine': 'Bakery',
 'grades': [{'date': datetime.datetime(2014, 3, 3, 0, 0),
             'grade': 'A',
             'score': 2},
            {'date': datetime.datetime(2013, 9, 11, 0, 0),
             'grade': 'A',
             'score': 6},
            {'date': datetime.datetime(2013, 1, 24, 0, 0),
             'grade': 'A',
             'score': 10},
            {'date': datetime.datetime(2011, 11, 23, 0, 0),
             'grade': 'A',
             'score': 9},
            {'date': datetime.datetime(2011, 3, 10, 0, 0),
             'grade': 'B',
             'score': 14}],
 'name': 'Morris Park Bake Shop',
 'restaurant_id': '30075445'}


In [69]:
db.restaurants.count() #Number of restaurants in the JSON file.

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use estimated_document_count or count_documents instead. Please note that $where must be replaced by $expr, $near must be replaced by $geoWithin with $center, and $nearSphere must be replaced by $geoWithin with $centerSphere
  """Entry point for launching an IPython kernel.


25359

# Inserting a new document:

In [70]:
db.restaurants.insert_one(
{
"address" : {
"street" : "2 Avenue",
"zipcode" : "10075",
"building" : "1480",
"coord" : [ -73.9557413, 40.7720266 ]
},
"borough" : "Manhattan",
"cuisine" : "Italian",
"grades" : [
{
"date" : datetime.datetime(2014, 10, 1, 0, 0, 0, 0),
    "grade" : "A",
"score" : 11
},
{
"date" : datetime.datetime(2014, 1, 16, 0, 0, 0, 0),
"grade" : "B",
"score" : 17
}
],
"name" : "Vella",
"restaurant_id" : "41704620"
}
)

In [71]:
db.restaurants.count()

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use estimated_document_count or count_documents instead. Please note that $where must be replaced by $expr, $near must be replaced by $geoWithin with $center, and $nearSphere must be replaced by $geoWithin with $centerSphere
  """Entry point for launching an IPython kernel.


25360

# Updating Top-level fields:

Update the Document relative to the Juni restaurant and set the cuisine to ‘American
(New)’ and the lastModified field in current date to True. Use the function find_one() before and after the update to
verify its execution.

In [72]:
queryF = { "name": "Juni" }
pprint(db.restaurants.find_one(queryF))
newValues={"$set":{"cuisine":"American(New)"}}
db.restaurants.update_one(queryF,newValues)
pprint(db.restaurants.find_one(queryF))

{'_id': ObjectId('5cc7253f3ff99b33afc7817e'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 31 Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score': 12},
            {'date': datetime.datetime(2013, 8, 5, 0, 0),
             'grade': 'A',
             'score': 5},
            {'date': datetime.datetime(2012, 6, 7, 0, 0),
             'grade': 'A',
             'score': 0}],
 'name': 'Juni',
 'restaurant_id': '41156888'}
{'_id': ObjectId('5cc7253f3ff99b33afc7817e'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 31 Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American(New)',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score': 12},
 

# Updating Embedded fields:

Update the street of the restaurant whose id is 41156888 with the new value "East
41st Street"

In [73]:
query2 = { "restaurant_id": "41156888" }
pprint(db.restaurants.find_one(query2))
newValues={"$set":{"address.street":"East 41st Street"}}
db.restaurants.update_one(query2,newValues)
pprint(db.restaurants.find_one(query2))

{'_id': ObjectId('5cc7253f3ff99b33afc7817e'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 31 Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American(New)',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score': 12},
            {'date': datetime.datetime(2013, 8, 5, 0, 0),
             'grade': 'A',
             'score': 5},
            {'date': datetime.datetime(2012, 6, 7, 0, 0),
             'grade': 'A',
             'score': 0}],
 'name': 'Juni',
 'restaurant_id': '41156888'}
{'_id': ObjectId('5cc7253f3ff99b33afc7817e'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 41st Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American(New)',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score':

# Updating multiple Documents:

Update all documents that have a zipcode field equal to "10016" and cuisine
field equal to “Other”. Set the cuisine field to "Category To Be Determined".

There is no field called "lastModified".

In [74]:
db.restaurants.find({"lastModified": {"$exists":True}}).count()

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


0

In [75]:
db.restaurants.find({"cuisine":"Other"}).count() #Number of restaurants with 'Other' value for the cuisine key.

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


1011

In [76]:
db.restaurants.update_many({"cuisine":"Other"},{"$set":{"cuisine":"Category To Be Determined"}}) #Updating multiple files.

In [77]:
db.restaurants.find({"cuisine":"Other"}).count() #We expect 0.

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


0

In [78]:
db.restaurants.find({"cuisine":"Category To Be Determined"}).count() # We expect 1011

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


1011

# Replacing a Document:

Update the Document for the restaurant with id : “41704620”. The modified document will only contain the _id field, name field, the address field. i.e. the document will not contain the restaurant_id, cuisine, grades, and the borough fields.

In [93]:
name=db.restaurants.find_one({"restaurant_id":"41704620"},{"name"})
address=db.restaurants.find_one({"restaurant_id":"41704620"},{"address"})
db.restaurants.replace_one({"restaurant_id":"41704620"},{
    "name":"ahmed",
    "address":"address"
})

In [95]:
db.restaurants.find_one({"restaurant_id":"41704620"})

In [96]:
pprint(db.restaurants.find_one({"restaurant_id":"41704620"}))

None
